qubes/vm/qubesvm.py

```diff
@@ -1446,60 +1495,32 @@ def get_power_state(self):
                 Libvirt's enum describing precise state of a domain.
         '''  # pylint: disable=too-many-return-statements
 
-        # don't try to define libvirt domain, if it isn't there, VM surely
-        # isn't running
-        # reason for this "if": allow vm.is_running() in PCI (or other
-        # device) extension while constructing libvirt XML
-        if self.app.vmm.offline_mode:
-            return 'Halted'
-        if self._libvirt_domain is None:
-            try:
-                self._libvirt_domain = self.app.vmm.libvirt_conn.lookupByUUID(
-                    self.uuid.bytes)
-            except libvirt.libvirtError as e:
-                if e.get_error_code() == libvirt.VIR_ERR_NO_DOMAIN:
-                    return 'Halted'
-                else:
-                    raise
-
-        libvirt_domain = self.libvirt_domain
-        if libvirt_domain is None:
-            return 'Halted'
-
-        try:
-            if libvirt_domain.isActive():
-                # pylint: disable=line-too-long
-                if libvirt_domain.state()[0] == libvirt.VIR_DOMAIN_PAUSED:
-                    return "Paused"
-                elif libvirt_domain.state()[0] == libvirt.VIR_DOMAIN_CRASHED:
-                    return "Crashed"
-                elif libvirt_domain.state()[0] == libvirt.VIR_DOMAIN_SHUTDOWN:
-                    return "Halting"
-                elif libvirt_domain.state()[0] == libvirt.VIR_DOMAIN_SHUTOFF:
-                    return "Dying"
-                elif libvirt_domain.state()[0] == libvirt.VIR_DOMAIN_PMSUSPENDED:  # nopep8
-                    return "Suspended"
-                else:
-                    if not self.is_fully_usable():
-                        return "Transient"
-
-                    return "Running"
-
-            return 'Halted'
-        except libvirt.libvirtError as e:
-            if e.get_error_code() == libvirt.VIR_ERR_NO_DOMAIN:
-                return 'Halted'
-            raise
-
-        assert False
+        state = self._libvirt_state
```

Comment:
The idea is to never call libvirtd for any read-only admin API call except for CPU/memory stats (and internal inspection).

It should theoretically not be fragile because it updates the state on every connection to libvirtd, on every lifecyle event received, and if the libvirt connection dies the close callback should be called, which triggers a reconnect, which triggers a state update.

Of course it's possible that the implementation is buggy

Without this doing qvm-ls requires at least one call to libvirtd per VM, which is probably going to take hundreds of milliseconds, and also any random "is_running" and similar call within qubes is going to cause I/O, which can also cause pervasive slowdown.